[quant][pt2e] Allow users to override train/eval behavior

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119091

Summary: This commit adds a util for PT2E quantization users
to call `model.train()` and `model.eval()` without error.
Instead, these will automatically call the equivalent
`move_exported_model_to_train/eval` for the user, which only
switch behavior for special ops like dropout and batchnorm.
This enables users to onboard to the PT2E flow more easily.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_exported_model_train_eval

Reviewers: jerryzh168, tugsbayasgalan, zhxchen17

Subscribers: jerryzh168, tugsbayasgalan, zhxchen17, supriyar

Differential Revision: [D53426636](https://our.internmc.facebook.com/intern/diff/D53426636)

andrewor14 committed:
[quant][pt2e] Allow users to override train/eval behavior

Summary: This commit adds a util for PT2E quantization users
to call `model.train()` and `model.eval()` without error.
Instead, these will automatically call the equivalent
`move_exported_model_to_train/eval` for the user, which only
switch behavior for special ops like dropout and batchnorm.
This enables users to onboard to the PT2E flow more easily.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_exported_model_train_eval_for_special_ops

Reviewers: jerryzh168, tugsbayasgalan, zhxchen17

Subscribers: jerryzh168, tugsbayasgalan, zhxchen17, supriyar

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/eval_utils.py:
```diff
@@ -137,3 +139,25 @@ def _move_exported_model_to_train(model: torch.fx.GraphModule):
+    def _train(self, mode: bool = True):
+        _move_exported_model_to_train(self)
+
+    def _eval(self, mode: bool = True):
```
`eval` doesn't have a `mode` argument actually: https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.eval

andrewor14 replied:
Good cath

jerryzh168 commented on torch/ao/quantization/__init__.py:
```diff
@@ -14,6 +14,7 @@
 from .stubs import *  # noqa: F403
 from .pt2e.eval_utils import _move_exported_model_to_eval as move_exported_model_to_eval
 from .pt2e.eval_utils import _move_exported_model_to_train as move_exported_model_to_train
+from .pt2e.eval_utils import _allow_exported_model_train_eval_for_special_ops as allow_exported_model_train_eval_for_special_ops
```
nit: maybe just define these as public methods in `eval_utils`? not sure what is the difference between this and defining as public methods directly

andrewor14 replied:
I kind of want to keep `eval_utils` a (semi) private namespace, since the file name is kind of TBD. Don't really want users to import from there directly. Do you think that's better than exposing this under top level toq?

jerryzh168 reviewed (approved):
looks great, thanks!

jerryzh168 commented on torch/ao/quantization/pt2e/eval_utils.py:
```diff
@@ -137,3 +139,25 @@ def _move_exported_model_to_train(model: torch.fx.GraphModule):
     _replace_batchnorm(model, train_to_eval=False)
     return model
+
+def _allow_exported_model_train_eval_for_special_ops(model: torch.fx.GraphModule):
```
nit: for naming, maybe just `allow_exported_model_train_eval`?

andrewor14 committed:
Update on "[quant][pt2e] Allow users to override train/eval behavior"


Summary: This commit adds a util for PT2E quantization users
to call `model.train()` and `model.eval()` without error.
Instead, these will automatically call the equivalent
`move_exported_model_to_train/eval` for the user, which only
switch behavior for special ops like dropout and batchnorm.
This enables users to onboard to the PT2E flow more easily.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_exported_model_train_eval_for_special_ops

Reviewers: jerryzh168, tugsbayasgalan, zhxchen17

Subscribers: jerryzh168, tugsbayasgalan, zhxchen17, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2e] Allow users to override train/eval behavior"


Summary: This commit adds a util for PT2E quantization users
to call `model.train()` and `model.eval()` without error.
Instead, these will automatically call the equivalent
`move_exported_model_to_train/eval` for the user, which only
switch behavior for special ops like dropout and batchnorm.
This enables users to onboard to the PT2E flow more easily.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_exported_model_train_eval_for_special_ops

Reviewers: jerryzh168, tugsbayasgalan, zhxchen17

Subscribers: jerryzh168, tugsbayasgalan, zhxchen17, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2e] Allow users to override train/eval behavior"


Summary: This commit adds a util for PT2E quantization users
to call `model.train()` and `model.eval()` without error.
Instead, these will automatically call the equivalent
`move_exported_model_to_train/eval` for the user, which only
switch behavior for special ops like dropout and batchnorm.
This enables users to onboard to the PT2E flow more easily.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_exported_model_train_eval

Reviewers: jerryzh168, tugsbayasgalan, zhxchen17

Subscribers: jerryzh168, tugsbayasgalan, zhxchen17, supriyar

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53426636).

andrewor14 commented:
@pytorchbot merge
